### PR TITLE
add invoked contract

### PR DESCRIPTION
**Pull request by Ashuaidehao**
Relate https://github.com/neo-project/neo-modules/pull/656.

### CODE REVIEW

**Comment by shargon** [comment]
@superboyiii tested?

**Comment by superboyiii** [comment]
> @superboyiii tested?

Give me more time.

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 1]
```diff
@@ -59,20 +59,24 @@ public void SerializeUnsigned(BinaryWriter writer)
             }
         }
 
-        private JObject GetInvokeResult(byte[] script, Signers signers = null)
+        private JObject GetInvokeResult(byte[] script, Signers signers = null, bool showDetails = false)
```
Comment (erikzhang) [inline src/RpcServer/RpcServer.SmartContract.cs:62]:
```suggestion
        private JObject GetInvokeResult(byte[] script, Signers signers = null, bool useDiagnostic = false)
```

**Comment by superboyiii** [comment]
@shargon 
invoke result of Neoverse Mint, can see `invokedcontracts` contains gas hash and neoverse hash.
1. invoked script hash(not contract), 2. GAS, 3. BlindBox, 4. StdLib, 5. LedgerContract, 6. ContractManagement
```
{
    "jsonrpc": "2.0",
    "id": 3,
    "result": {
        "script": "DANuZnQCAMqaOwwUQG9HkCtsPwiAbqyM+fMqDVxetrMMFMCY5KzwsgkN0W7LPVjeQmqHFYUbFMAfDAh0cmFuc2ZlcgwUz3bii9AGLEpHjuNVYQETGfPPpNJBYn1bUg==",
        "state": "HALT",
        "gasconsumed": "21513739",
        "exception": null,
        "invokedcontracts": [
            "0x7411589817d51cfb6ee7d322a1f2dee8c5613185",
            "0xd2a4cff31913016155e38e474a2c06d08be276cf",
            "0xb3b65e5c0d2af3f98cac6e80083f6c2b90476f40",
            "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0",
            "0xda65b600f7124ce6c79950c1772a36403104f2be",
            "0xfffdc93764dbaddd97c48f252a53ea4643faa3fd"
        ],
        "stack": [
            {
                "type": "Boolean",
                "value": true
            }
        ],
        "tx": "ACaP5i4LRkgBAAAAAKDiAQAAAAAA4R8KAAHAmOSs8LIJDdFuyz1Y3kJqhxWFG4AAXgwDbmZ0AgDKmjsMFEBvR5ArbD8IgG6sjPnzKg1cXrazDBTAmOSs8LIJDdFuyz1Y3kJqhxWFGxTAHwwIdHJhbnNmZXIMFM924ovQBixKR47jVWEBExnzz6TSQWJ9W1IBQgxAiZrW3UeII/mf9oKMTWv3p+T4Q8bAJySMjNcRYHAMvMK2ZxkgVUMvYnLdd7uu7ncu23M0UAnme6It/9QYCMH4TSgMIQKNWcem5o/HrWiwhxgZW8nce5BzQ4To6abGD/mj+1t7PEFW57Mn"
    }
}

```

**Review by shargon — APPROVED** [review]

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 2]
```diff
@@ -59,20 +59,27 @@ public void SerializeUnsigned(BinaryWriter writer)
...
             JObject json = new();
             json["script"] = Convert.ToBase64String(script);
             json["state"] = engine.State;
             json["gasconsumed"] = engine.GasConsumed.ToString();
             json["exception"] = GetExceptionMessage(engine.FaultException);
+            if (useDiagnostic)
+            {
+                json["diagnostics"] = new JObject()
+                {
+                    ["invokedcontracts"] = engine.Diagnostic.InvocationTree.GetItems().Distinct().Select(v => (JString)v.ToString()).ToArray()
```
Comment (erikzhang) [inline src/RpcServer/RpcServer.SmartContract.cs:80]: Returns the `InvocationTree`?

Comment (shargon) [reply src/RpcServer/RpcServer.SmartContract.cs:80]: Also we can return the storage changes?

Comment (Ashuaidehao) [reply src/RpcServer/RpcServer.SmartContract.cs:80]: Rename？

Comment (erikzhang) [reply src/RpcServer/RpcServer.SmartContract.cs:80]: It is not a simple renaming, but also storing the tree structure into an array.

Comment (Ashuaidehao) [reply src/RpcServer/RpcServer.SmartContract.cs:80]: `InvocationTree` seems cannot traverse tree structure outside now. Need some more support.

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 3]
```diff
@@ -59,20 +59,27 @@ public void SerializeUnsigned(BinaryWriter writer)
...
-            using ApplicationEngine engine = ApplicationEngine.Run(script, system.StoreView, container: tx, settings: system.Settings, gas: settings.MaxGasInvoke);
+            using ApplicationEngine engine = ApplicationEngine.Run(script, system.StoreView, container: tx, settings: system.Settings, gas: settings.MaxGasInvoke, diagnostic: useDiagnostic ? new Diagnostic() : null);
             JObject json = new();
             json["script"] = Convert.ToBase64String(script);
             json["state"] = engine.State;
             json["gasconsumed"] = engine.GasConsumed.ToString();
             json["exception"] = GetExceptionMessage(engine.FaultException);
+            if (useDiagnostic)
+            {
+                json["diagnostics"] = new JObject()
```
Comment (roman-khimov) [inline src/RpcServer/RpcServer.SmartContract.cs:78]: BTW, do we need this wrapper structure?

Comment (erikzhang) [reply src/RpcServer/RpcServer.SmartContract.cs:78]: Maybe we will add more data to it.

**Comment by roman-khimov** [comment]
Going back to the https://github.com/neo-project/neo-modules/pull/656#discussion_r729934698 list, can we also add notifications as something always returned for `invoke*` calls? The node has this data already, but we don't give it back to the user and it also might be important for the receiver to ensure that side-effects produced by transaction are the ones intended.

**Comment by shargon** [comment]
@Ashuaidehao `NeoFS.API` must be updated

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 4]
```diff
@@ -88,6 +105,27 @@ private JObject GetInvokeResult(byte[] script, Signers signers = null)
...
+            }
+            return json;
+        }
+
+        private static JObject ToJson(System.Collections.Generic.IEnumerable<DataCache.Trackable> changes)
+        {
+            JArray json = new();
+            foreach (var entry in changes)
+            {
+                json[entry.Key.ToArray().ToHexString()] = Convert.ToBase64String(entry.Item.Value);
```
Comment (roman-khimov) [inline src/RpcServer/RpcServer.SmartContract.cs:124]: This differs from `StatesDumper` format, do we have any good reason for this difference?

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 5]
```diff
@@ -74,6 +74,14 @@ private JObject GetInvokeResult(byte[] script, Signers signers = null, bool useD
             json["state"] = engine.State;
             json["gasconsumed"] = engine.GasConsumed.ToString();
             json["exception"] = GetExceptionMessage(engine.FaultException);
+            json["notifications"] = new JArray(engine.Notifications.Select(n =>
+              {
+                  var obj = new JObject();
+                  obj["event"] = n.EventName;
+                  obj["contract"] = n.ScriptHash.ToString();
+                  obj["parameters"] = ToJson(n.State, settings.MaxIteratorResultItems);
```
Comment (roman-khimov) [inline src/RpcServer/RpcServer.SmartContract.cs:82]: Differs from ApplicationLog JSONization of the same data.

File: src/RpcServer/RpcServer.SmartContract.cs (hunk 4 above)
Comment (shargon) [reply src/RpcServer/RpcServer.SmartContract.cs]: None, fixed :)

**Review by shargon — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Comment by superboyiii** [comment]
Test Failed:
Current ["storagechanges"]  will overwrite ["invokedcontracts"] in json["diagnostics"]

**Comment by superboyiii** [comment]
Test Pass
Now it works well.
```
{
    "jsonrpc": "2.0",
    "id": 3,
    "result": {
        "script": "DA9CbGluZCBCb3ggIzE3NTERwB8MCHVuQm94aW5nDBR/jtStf6tG7M4EV8/3LtzjPQni2UFifVtS",
        "state": "HALT",
        "gasconsumed": "7757705",
        "exception": null,
        "notifications": [
            {
                "eventname": "Transfer",
                "contract": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                "state": {
                    "type": "Array",
                    "value": [
                        {
                            "type": "ByteString",
                            "value": "wJjkrPCyCQ3Rbss9WN5CaocVhRs="
                        },
                        {
                            "type": "Any"
                        },
                        {
                            "type": "Integer",
                            "value": "1"
                        },
                        {
                            "type": "ByteString",
                            "value": "QmxpbmQgQm94ICMxNzUx"
                        }
                    ]
                }
            },
            {
                "eventname": "Transfer",
                "contract": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                "state": {
                    "type": "Array",
                    "value": [
                        {
                            "type": "Any"
                        },
                        {
                            "type": "ByteString",
                            "value": "wJjkrPCyCQ3Rbss9WN5CaocVhRs="
                        },
                        {
                            "type": "Integer",
                            "value": "1"
                        },
                        {
                            "type": "ByteString",
                            "value": "RnJhZ21lbnQgQSAjNDU2"
                        }
                    ]
                }
            }
        ],
        "diagnostics": {
            "invokedcontracts": {
                "hash": "0x18f4ea90a5612a7fdd6f7ef0928e6d3686d2d48a",
                "call": [
                    {
                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                        "call": [
                            {
                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                            },
                            {
                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                "call": [
                                    {
                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                        "call": [
                                            {
                                                "hash": "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0"
                                            }
                                        ]
                                    },
                                    {
                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                    },
                                    {
                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                        "call": [
                                            {
                                                "hash": "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0"
                                            }
                                        ]
                                    },
                                    {
                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                        "call": [
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                                "call": [
                                                    {
                                                        "hash": "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0"
                                                    }
                                                ]
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                                "call": [
                                                    {
                                                        "hash": "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0"
                                                    }
                                                ]
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                                "call": [
                                                    {
                                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                                        "call": [
                                                            {
                                                                "hash": "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0"
                                                            }
                                                        ]
                                                    },
                                                    {
                                                        "hash": "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0"
                                                    }
                                                ]
                                            }
                                        ]
                                    },
                                    {
                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                        "call": [
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                                "call": [
                                                    {
                                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                                    },
                                                    {
                                                        "hash": "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0"
                                                    }
                                                ]
                                            }
                                        ]
                                    },
                                    {
                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                        "call": [
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                            }
                                        ]
                                    },
                                    {
                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                        "call": [
                                            {
                                                "hash": "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0"
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                                "call": [
                                                    {
                                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                                    }
                                                ]
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                            }
                                        ]
                                    },
                                    {
                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                        "call": [
                                            {
                                                "hash": "0xacce6fd80d44e1796aa0c2c625e9e4e0ce39efc0"
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                                "call": [
                                                    {
                                                        "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                                    }
                                                ]
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f"
                                            },
                                            {
                                                "hash": "0xd9e2093de3dc2ef7cf5704ceec46ab7fadd48e7f",
                                                "call": [
                                                    {
                                                        "hash": "0xfffdc93764dbaddd97c48f252a53ea4643faa3fd"
                                                    }
                                                ]
                                            }
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                ]
            },
            "storagechanges": [
                {
                    "state": "Added",
                    "key": "aAAAABMAyAE=",
                    "value": "6Ao="
                },
                {
                    "state": "Changed",
                    "key": "aAAAABIw",
                    "value": "5wo="
                },
                {
                    "state": "Changed",
                    "key": "aAAAABFjb3VudGVy",
                    "value": "xQA="
                },
                {
                    "state": "Deleted",
                    "key": "aAAAAANCbGluZCBCb3ggIzE3NTE=",
                    "value": "QAcoFMCY5KzwsgkN0W7LPVjeQmqHFYUbKA9CbGluZCBCb3ggIzE3NTEhACEAIQLXBigcaHR0cHM6Ly9uZW8ub3JnL0JsaW5kQm94LnBuZygcaHR0cHM6Ly9uZW8ub3JnL0JsaW5kQm94Lm1wNA=="
                },
                {
                    "state": "Changed",
                    "key": "aAAAAAHAmOSs8LIJDdFuyz1Y3kJqhxWFGw==",
                    "value": "UAE="
                },
                {
                    "state": "Deleted",
                    "key": "aAAAAATAmOSs8LIJDdFuyz1Y3kJqhxWFG0JsaW5kIEJveCAjMTc1MQ==",
                    "value": ""
                },
                {
                    "state": "Changed",
                    "key": "aAAAAAA=",
                    "value": "4g0="
                },
                {
                    "state": "Added",
                    "key": "aAAAAANGcmFnbWVudCBBICM0NTY=",
                    "value": "QAcoFMCY5KzwsgkN0W7LPVjeQmqHFYUbKA9GcmFnbWVudCBBICM0NTYhAQEhACECyAEoHWh0dHBzOi8vbmVvLm9yZy9GcmFnbWVudEEucG5nKB1odHRwczovL25lby5vcmcvRnJhZ21lbnRBLm1wNA=="
                },
                {
                    "state": "Added",
                    "key": "aAAAAATAmOSs8LIJDdFuyz1Y3kJqhxWFG0ZyYWdtZW50IEEgIzQ1Ng==",
                    "value": ""
                }
            ]
        },
        "stack": [
            {
                "type": "Boolean",
                "value": true
            }
        ]
    }
}
```

**Review by shargon — DISMISSED** [review]

**Review by shargon — APPROVED** [review]

**Pull request merged**